File: go/20.valid-parentheses.go
/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */
package main

import (
	"fmt"
	"container/list"
)
func isValid(s string) bool {
	stack := list.New()
	memo := make(map[rune]rune)
	memo[')'] = '('
	memo[']'] = '['
	memo['}'] = '{'
    for _, c := range s {
		if c == '(' || c == '{' || c == '[' {
			stack.PushBack(c)
		} else {
			if stack.Len() > 0 {
				if memo[c] != stack.Remove(stack.Back()).(rune) {
					return false
				}
			} else {
				return false
			}
		}
	}

	if stack.Len() == 0 { return true } else { return false }
}

// func main() {
// 	s := "["
// 	fmt.Println(isValid(s))
// }

File: rust/20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

use std::collections::HashMap;

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut dict:HashMap<&char, &char> = HashMap::new();
        dict.insert(&']', &'[');
        dict.insert(&'}', &'{');
        dict.insert(&')', &'(');

        let mut vec = Vec::new();
        for c in s.chars() {
            match c {
                '(' | '{' | '[' => { vec.push(c); },
                _ => {
                    let a = vec.pop();
                    let v = dict.get(&c);
                        
                    match a {
                        Some(x) => {
                            if x != **v.unwrap() {
                                return false;
                            }
                        },
                        None => return false,
                    }
                },
            }
        }

        vec.len() == 0
    }
}

